### ISSUE TITLE
`lia` segfaults

### ISSUE THREAD

**Issue by robbertkrebbers** [coq/coq#10957]
```coq
Require Import Psatz Fin.

Fixpoint fin_to_nat {n} (i : Fin.t n) : nat :=
  match i with F1 => 0 | FS i => S (fin_to_nat i) end.

Lemma foo (n : nat) (i : Fin.t n) :
  0 = S (fin_to_nat i) ->
  F1 = FS i.
Proof. lia. (* segmentation fault *)
```

#### Coq Version

Using Coq master 6ed3b02af77313d62ec868b4a88a208a9003857d

```
$ coqtop -v
The Coq Proof Assistant, version 8.11+alpha (October 2019)
compiled on Oct 25 2019 9:04:16 with OCaml 4.05.0
```

**Comment by robbertkrebbers** [comment]
Yes, it works with 8.10.1

**Comment by robbertkrebbers** [comment]
I'm using:

```
$ ocaml --version
The OCaml toplevel, version 4.05.0
```

**Comment by robbertkrebbers** [comment]
This issue is kind of a blocker for me. Debian testing ships OCaml 4.05.0, so it's impossible to build Coq master now.

**Comment by ppedrot** [comment]
Still cannot reproduce on the same commit and using the 4.05.0 OCaml compiler shipped with Debian. Did you try from a fresh install?

**Comment by ppedrot** [comment]
Relevant for this kind of problem:
```
pm@ouranos:~/sources/coq$ ulimit -s
8192
```

**Comment by robbertkrebbers** [comment]
It appears to have something to do with my `.lia.cache`. I removed that file, and now `lia` works again without segfaulting.

I still think this is a bug, since `coqc` should never segfault.

**Comment by silene** [comment]
"never" is too strong a word. As @ppedrot mentioned, a resource shortage (either stack or heap) is a perfectly fine segfault.

**Comment by ppedrot** [comment]
@robbertkrebbers This is most likely caused by the fact that lia uses the OCaml marshalling functions, which are type-unsound. Changing the title accordingly.

**Comment by ejgallego** [comment]
Indeed a different serialization method could help, but as a short term fix a magic number could help avoid these kind of problems.

**Comment by ppedrot** [comment]
I think it's fairly easy to remove the need to OCaml marshaller, I'll write a quick patch.

**Comment by maximedenes** [comment]
@ppedrot any update?

**Comment by ppedrot** [comment]
@maximedenes nope, it's simple but tedious to write the marshallers and I am lazy.

**Comment by maximedenes** [comment]
@ppedrot do you need to be forced? ;)
